be/src/exec/spill/input_stream.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/spill/input_stream.h"
+
+#include "runtime/sorted_chunks_merger.h"
+#include "util/blocking_queue.hpp"
+#include "util/defer_op.h"
+
+namespace starrocks {
+namespace spill {
+
+static const int chunk_buffer_max_size = 2;
+
+class BufferedInputStream : public InputStream {
+public:
+    BufferedInputStream(int capacity, InputStreamPtr stream) : _capacity(capacity), _input_stream(stream) {}
+    ~BufferedInputStream() override = default;
+
+    bool is_buffer_full() { return _chunk_buffer.get_size() == _capacity || eof(); }
+    bool has_chunk() { return !_chunk_buffer.empty() || eof(); }
+
+    StatusOr<ChunkUniquePtr> get_next(FormatterContext& ctx) override;
+    bool is_ready() override { return has_chunk(); }
+    void close() override {}
+
+    bool enable_prefetch() const override { return true; }
+
+    Status prefetch(FormatterContext& ctx) override;
+
+    StatusOr<ChunkUniquePtr> read_from_buffer();
+
+private:
+    bool _acquire() {
+        bool expected = false;
+        return _is_prefetching.compare_exchange_strong(expected, true);
+    }
+    void _release() { _is_prefetching = false; }
+
+private:
+    int _capacity;
+    InputStreamPtr _input_stream;
+    UnboundedBlockingQueue<ChunkUniquePtr> _chunk_buffer;
+    std::atomic_bool _is_prefetching = false;
+};
+
+StatusOr<ChunkUniquePtr> BufferedInputStream::read_from_buffer() {
+    if (_chunk_buffer.empty()) {
+        CHECK(eof());
+        return Status::EndOfFile("end of stream");
+    }
+    ChunkUniquePtr res;
+    CHECK(_chunk_buffer.try_get(&res));
+    return res;
+}
+
+StatusOr<ChunkUniquePtr> BufferedInputStream::get_next(FormatterContext& ctx) {
+    if (has_chunk()) {
+        return read_from_buffer();
+    }
+    CHECK(!_is_prefetching);
+    return _input_stream->get_next(ctx);
+}
+
+Status BufferedInputStream::prefetch(FormatterContext& ctx) {
+    if (is_buffer_full()) {
+        return Status::OK();
+    }
+    // concurrent prefetch is not allowed, should call _acquire and _release before and after prefetch
+    // to ensure that it doesn't happen.
+    if (!_acquire()) {
```

Comment:
acquire function at here may not as expected? suggest move it to SpilledInputStream